src/s_inter.c

```diff
@@ -74,11 +74,64 @@ that didn't really belong anywhere. */
 #include "pthread.h"
 #endif
 
+#ifdef THREADED_IO
+#include "ringbuffer.h"
+
+#if __STDC_VERSION__ >= 201112L // use stdatomic if C11 is available
+  #include <stdatomic.h>
+  #define SYNC_FETCH(ptr) atomic_fetch_or((_Atomic int *)ptr, 0)
+  #define SYNC_COMPARE_AND_SWAP(ptr, oldval, newval) \
+          atomic_compare_exchange_strong((_Atomic int *)ptr, &oldval, newval)
+  #define SYNC_STORE(ptr, newval) atomic_store((_Atomic int *)ptr, newval)
+  #define t_atomic_int int
+#else // use platform specfics
+  #ifdef __APPLE__ // apple atomics
+    #include <libkern/OSAtomic.h>
+    #define SYNC_FETCH(ptr) OSAtomicOr32Barrier(0, (volatile uint32_t *)ptr)
+    #define SYNC_COMPARE_AND_SWAP(ptr, oldval, newval) \
+            OSAtomicCompareAndSwap32Barrier(oldval, newval, ptr)
+    #define t_atomic_int int32_t
+  #elif defined(_WIN32) || defined(_WIN64) // win api atomics
+    #include <windows.h>
+    #define SYNC_FETCH(ptr) InterlockedOr(ptr, 0)
+    #define SYNC_COMPARE_AND_SWAP(ptr, oldval, newval) \
+            InterlockedCompareExchange(ptr, oldval, newval)
+    #define t_atomic_int LONG
+  #else // gcc atomics
+    #define SYNC_FETCH(ptr) __sync_fetch_and_or(ptr, 0)
+    #define SYNC_COMPARE_AND_SWAP(ptr, oldval, newval) \
+            __sync_val_compare_and_swap(ptr, oldval, newval)
+    #define t_atomic_int int
+  #endif
+  #define SYNC_STORE(ptr, newval) do { t_atomic_int* p = ptr; SYNC_COMPARE_AND_SWAP(p, *p, newval); } while (0);
```

Comment:
Nitpick: since we now have a well-define atomic int type, this could simply be
`#define SYNC_STORE(ptr, newval) SYNC_COMPARE_AND_SWAP(ptr, *ptr, newval)`

Reply:
The intermediate `p` is to avoid using the macro parameter twice because it could have side effects:
say that one calls `SYNC_STORE(myPtr++)`, then `myPtr` would be incremented twice in the macro expansion.

Reply:
> because it could have side effects:

That's true! But doesn't this also apply to `SYNC_COMPARE_AND_SWAP`? If you want to fix this, you would rather have to do it there.

On the other hand, since we now have a well-defined atomic int type, those macros could be proper functions. This would make them typesafe and prevent double evaluation. The compiler will inline them anyway.

By the way, since those macros are essentially copy pasted from `ringbuffer.c`, they could actually go into `ringbuffer.h`, which is included by both `ringbuffer.c` and `sys_inter.c`. However, if we choose function over macros and put them in a header, they need to be `static` to avoid linker errors.

Reply:
(Sorry again for all the nitpicking. I really like this PR and hope it gets merged eventually!)

Reply:
But as you've written in your other comment, I can clean this up myself later (just have to remember it :-), so don't bother.

Reply:
> But doesn't this also apply to SYNC_COMPARE_AND_SWAP? If you want to fix this, you would rather have to do it there.

I don't think so: it only uses each parameter exactly once.

> By the way, since those macros are essentially copy pasted from ringbuffer.c, they could actually go into ringbuffer.h

I figured at some point they should all go in a `s_atomic.h` header.

Reply:
> I don't think so: it only uses each parameter exactly once.

:facepalm: you're completely right!

> I figured at some point they should all go in a s_atomic.h header.

+1